internal/worker/upgradeservices/worker.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package upgradeservices
+
+import (
+	"github.com/juju/worker/v4"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/changestream"
+	coreerrors "github.com/juju/juju/core/errors"
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/internal/errors"
+	internalservices "github.com/juju/juju/internal/services"
+)
+
+// Config is used to supply the required information for constructing a new
+// upgrade services worker.
+type Config struct {
+	// DBGetter supplies WatchableDB implementations by namespace.
+	DBGetter changestream.WatchableDBGetter
+
+	// Logger is to be used by the upgrade services.
+	Logger logger.Logger
+
+	// NewUpgradeServicesGetter provides a new
+	// [internalservices.UpgradeServicesGetter] for getting the upgrade
+	// services.
+	NewUpgradeServicesGetter UpgradeServicesGetterFn
+
+	// NewUpgradeServices provides a way to construct a new
+	// [internalservices.UpgradeServices].
+	NewUpgradeServices UpgradeServicesFn
+}
+
+// servicesWorker is the long running worker in the controller supplying
+// [internalservices.UpgradeServices] implementation.
+type servicesWorker struct {
+	tomb tomb.Tomb
+
+	servicesGetter internalservices.UpgradeServicesGetter
+}
+
+// upgradeServicesGetter provides a worker based implementation of
+// [internalservices.UpgradeServicesGetter].
+type upgradeServicesGetter struct {
+	dbGetter           changestream.WatchableDBGetter
+	logger             logger.Logger
+	newUpgradeServices UpgradeServicesFn
+}
+
+// UpgradeServicesGetterFn describes a func type capable of returning a new
+// [internalservices.UpgradeServicesGetter]
+type UpgradeServicesGetterFn func(
+	UpgradeServicesFn, changestream.WatchableDBGetter, logger.Logger,
+) internalservices.UpgradeServicesGetter
+
+// UpgradeServicesFn describes a func type capable of returning a new
+// [internalservices.UpgradeServices].
+type UpgradeServicesFn func(
+	changestream.WatchableDBGetter, logger.Logger,
+) internalservices.UpgradeServices
+
+// Kill kills this [servicesWorker].
```

Comment:
Either drop the stuttering:
```suggestion
// Kill this [servicesWorker].
```

or rephrase:
```suggestion
// Kill marks this [servicesWorker] as dying.
```

Reply:
Done & Done